### DISCUSSION TITLE
Using useSWR in components deep down the React DOM

### DISCUSSION THREAD

**Discussion by dMaisnam** [vercel/swr#1475] [Q&A]
Hello everyone. I just wanted to get clear about how do I fetch data from within a component deep down in the react DOM? Shouldn't I be able to fetch data using `useSWR` from anywhere in my react components by defining the `SWRConfig` in the `App.jsx` similar to how context work with `React.useContext()`.
My code in `src/App.jsx`
```
const fetcher = (url) => axios.get(url).then(res => res.data);
// const { data, error } = useSWR(API_URL, fetcher);
  
// if (error) return <div>Error</div>
// if (!data) return <div>Loading...</div>
  return (
    <SWRConfig value={{fetcher}}>
      <BrowserRouter basename="/">
        <Switch>
          <Route exact path="/" component={Homepage} />
          <Route exact path="/blog" component={BlogPage} /> 
          <Route path="*" component={NotFound} />
        </Switch>
      </BrowserRouter>
    </SWRConfig>
  )
```
In `src/components/BlogPage.jsx`
```
...
export default function BlogPage() {
  const { data, error } = useSWR(API_URL);
  
  if (error) return <div>Error</div>
  if (!data) return <div>Loading...</div>

  return <Layout>
    { data.blogs.map(blog) => {
        <article key={blog.id}>
          <Article blog={blog} />
        </article>
      }
    }
  </Layout>
}
```
When I do the data fetching logic in my `App` component, it works as expected. But when I fetch data from within my `BlogPage` component, an ErrorBoundary I set up in `src/index.jsx` is triggered which brings me to 2 questions I want to ask.

First, how do I fetch data from any component in the DOM using swr?
Second, even if data fetching fails, shouldn't the error described in `if (error) return <div>Error</div>` be triggered?
P.S I am working with react 17.0.2, react-router-dom 5.2.0 & swr 1.0.0 if it is at all relevant to my problem. 
Any help is much appreciated. Thank you for your time.

**Comment by shuding** [comment]
Please read https://swr.vercel.app/docs/getting-started#make-it-reusable, we recommend to make custom hooks and directly use it inside your components.

**Comment by shuding** [comment]
> Second, even if data fetching fails, shouldn't the error described in if (error) return <div>Error</div> be triggered?

Yes, it should return the "Error" text. If you're seeing something unexpected, could you please create a minimal reproduction on CodeSandbox?

**Reply by dMaisnam** [reply]
https://codesandbox.io/s/react-swr-example-lcbpr?file=/src/pages/BlogPage.jsx
I am surprised it is working fine in codesandbox
What is confusing me  in my case is that, if I perform the `useSWR` data fetch for the `/blog` route in both the `App.jsx` (i.e, uncommenting line 14 to 20 in `src/App.jsx`) & in `src/pages/BlogPage.jsx`, it works, but if I do the data fetch from within the `src/pages/BlogPage.jsx` component only, the `ErrorBoundary` set up in `src/index.js` is always triggered instead of the swr error

**Reply by dMaisnam** [reply]
I don't know if this was the cause but replacing the `<FetchLoading />` and `<Error />` components with JSX elements like `<div>Loading</div>` and `<div>Error fetching data</div>` fixed the error.
I am not sure if importing and using the `<Loading />` and `<Error />` components in `BlogPage` caused the issue. Weird because it works perfectly fine in the codesandbox
